ApiReview/API.Polly.Core/NoDocs/Polly/CompositeStrategyBuilderBase.cs

```diff
@@ -0,0 +1,27 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Polly;
+
+public abstract class CompositeStrategyBuilderBase
+{
+    public string? Name { get; set; }
+    public string? InstanceName { get; set; }
+    public ResilienceProperties Properties { get; }
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public Action<IList<ResilienceStrategy>>? OnCreatingStrategy { get; set; }
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public DiagnosticSource? DiagnosticSource { get; set; }
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    [Required]
+    public Func<double> Randomizer { get; set; }
```

Comment:
When the library creates a new instance of a strategy, it passes some of the properties of this type to the strategy factory methods which then uses them to find the types it needs.

so far, this type is used to pass the information about 3 injectable implementations plus ResilienceProperties is used for IServiceProvider injections.

can we generalize the approach here? 
Why don't we use IServiceProvider in the first place to resolve all these customizable pieces?

Reply:
> so far, this type is used to pass the information about 3 injectable implementations plus ResilienceProperties is used for IServiceProvider injections.

I agree this can be confusing. 

> can we generalize the approach here?

I can make the following changes so the strategies will have only one way of configuring:

- `Randomizer` will not be exposed on builder as it's not common enough. Instead, it will be moved to `RetryStrategyOptions`
- The `ResilienceProperties` will be dropped. This was only "academical API" aimed to attach custom attributes to the builder that currently has no real use-case. We can re-introduce later on customer ask.
- `Telemetry` and `TimeProvider` will be provided by Polly infra when creating strategies, however, not directly modifiable.

> Why don't we use IServiceProvider in the first place to resolve all these customizable pieces?

I am hesitant to introduce `IServiceProvider` because it hides the dependencies. `Polly.Core` is standalone package and strategies should be explicit about their dependencies. The should express their dependencies using the strategy options. 

Consumer can can still resolve these dependencies from DI and just pass them to options. Of course, if there is a real use-case for `IServiceProvider` (now or in the future) I am not against exposing it.

Reply:
The builder has now only 2 properties that end-consumer can set:

- `Name` 
- `InstanceName`